cmd/juju/cloud/add_test.go

```diff
@@ -7,172 +7,256 @@ import (
 	"bytes"
 	"io"
 	"io/ioutil"
-	"path/filepath"
 	"strings"
 
 	"github.com/juju/cmd"
+	jujutesting "github.com/juju/testing"
 	jc "github.com/juju/testing/checkers"
 	"github.com/pkg/errors"
 	gc "gopkg.in/check.v1"
 
+	cloudfile "github.com/juju/juju/cloud"
 	"github.com/juju/juju/cmd/juju/cloud"
-	"github.com/juju/juju/juju/osenv"
 	"github.com/juju/juju/testing"
 )
 
 type addSuite struct {
-	testing.FakeJujuXDGDataHomeSuite
+	jujutesting.IsolationSuite
 }
 
 var _ = gc.Suite(&addSuite{})
 
+func newFakeCloudQueryer() *fakeCloudQueryer {
+	return &fakeCloudQueryer{CallMocker: jujutesting.NewCallMocker()}
+}
+
+type fakeCloudQueryer struct {
+	cloud.CloudQueryer
+	*jujutesting.CallMocker
+}
+
+func (f *fakeCloudQueryer) ParseCloudMetadataFile(path string) (map[string]cloudfile.Cloud, error) {
+	results := f.MethodCall(f, "ParseCloudMetadataFile", path)
+	return results[0].(map[string]cloudfile.Cloud), typeAssertError(results[1])
+}
+
+func (f *fakeCloudQueryer) PublicCloudMetadata(searchPaths ...string) (result map[string]cloudfile.Cloud, fallbackUsed bool, _ error) {
+	results := f.MethodCall(f, "PublicCloudMetadata", searchPaths)
+	return results[0].(map[string]cloudfile.Cloud), results[1].(bool), typeAssertError(results[2])
+}
+
+func (f *fakeCloudQueryer) PersonalCloudMetadata() (map[string]cloudfile.Cloud, error) {
+	results := f.MethodCall(f, "PersonalCloudMetadata")
+	return results[0].(map[string]cloudfile.Cloud), typeAssertError(results[1])
+}
+
+func (f *fakeCloudQueryer) WritePersonalCloudMetadata(cloudsMap map[string]cloudfile.Cloud) error {
+	results := f.MethodCall(f, "WritePersonalCloudMetadata", cloudsMap)
+	return typeAssertError(results[0])
+}
+
+func (f *fakeCloudQueryer) ParseOneCloud(data []byte) (cloudfile.Cloud, error) {
+	results := f.MethodCall(f, "ParseOneCloud", data)
+	return results[0].(cloudfile.Cloud), typeAssertError(results[1])
+}
+
+func typeAssertError(err interface{}) error {
```

Comment:
could we just call this "toError"?  Would be clearer, IMO.  Also, an explanation of why you can't just do results[x].(error) would be nice, because it's not immediately clear why that wouldn't work (after thinking about it for a minute, I got it, but I prefer not to make people think, if possible ;)